### PR TITLE
updates to vyos_config module arguments

### PR DESCRIPTION
**Pull request by privateip**
- arguments for vyos_config for 2.2 are now complete
- adds loading config file from disk (src argument)
- removes unsupported rollback argument
- changes update_config to update with options merge or check
- changes backup_config to backup
- add state argument for state of configuration file
- adds backup argument to backup current configuration
- adds save argument to control if active config is saved to disk
- adds comment argument for setting commit comment
- adds match argument to control configuraiton match

Tested with VyOS 1.7

### CODE REVIEW

**Comment by privateip** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;GlyP3Sdx&#x2F;65-vyos-config"> vyos_config</a>

**Comment by gregdek** [comment]
Thanks again to @privateip. This module is going into community review.

We notice that you are one of the maintainers (@privateip, @gundalow) of this module, so when you think it's ready, please comment "shipit" and we will consider it for merging.

[This message brought to you by your friendly Ansibull-bot.]

File: network/vyos/vyos_config.py [hunk 1]
```diff
@@ -99,30 +142,43 @@
...
+]
 
 
 def invoke(name, *args, **kwargs):
     func = globals().get(name)
     if func:
         return func(*args, **kwargs)
 
+def check_args(module, warnings):
+    if module.params['save'] and module.params['udpate'] == 'check':
```
Comment (gundalow) [inline network/vyos/vyos_config.py]: typo of `update`

File: network/vyos/vyos_config.py [hunk 2]
```diff
@@ -99,30 +142,43 @@
...
+        warnings.append('The configuration will not be saved when update '
```
Comment (gundalow) [inline network/vyos/vyos_config.py:172]: Should this be a warning or fatal error?

Just wondering if this warning could get lost in the sea of output of ansible-playbook

File: network/vyos/vyos_config.py [hunk 3]
```diff
@@ -133,74 +189,141 @@ def config_to_commands(config):
...
+        elif item.startswith('set') and item not in config:
+            updates.append(line)
+
+        elif item.startswith('delete'):
+            if not config:
+                updates.append(line)
+            else:
+                item = re.sub(r'delete', 'set', item)
+                for entry in config:
+                    if entry.startswith(item) and not line in visited:
```
Comment (gundalow) [inline network/vyos/vyos_config.py]: vyos_config.py:227:51: E713 test for membership should be 'not in'

File: network/vyos/vyos_config.py [hunk 4]
```diff
@@ -39,46 +39,89 @@
...
+        always loaded.
     required: false
-    default: null
-  update_config:
+    default: line
+    choices: ['line', 'none']
+  update:
+    description:
+      - The C(update) argument controls the method used to update the
+        remote device configuraiton.  This argument accepts two valid
```
Comment (gundalow) [inline network/vyos/vyos_config.py]: configuration

File: network/vyos/vyos_config.py [hunk 5]
```diff
@@ -39,46 +39,89 @@
...
+        when the configuration is committed.  If the configuration is
+        not changed or committed, this argument is ignored.
+    required: false
+    default: 'configured by vyos_config'
+  config:
+    description:
+      - The C(config) argument specifies the base configuration to use
+        to compare against the desired configuration.  If this value
+        is not specified, the module will automatically retrieve the
+        current active configuraiton from the remote device.
```
Comment (gundalow) [inline network/vyos/vyos_config.py]: configuration

File: network/vyos/vyos_config.py [hunk 6]
```diff
@@ -39,46 +39,89 @@
...
+    required: false
+    default: null
+  save:
+    description:
+      - The C(save) arugment controls whether or not changes made
```
Comment (gundalow) [inline network/vyos/vyos_config.py]: argument

File: network/vyos/vyos_config.py [hunk 7]
```diff
@@ -133,74 +189,141 @@ def config_to_commands(config):
         commands = str(candidate).split('\n')
     return commands
```
Comment (gundalow) [inline network/vyos/vyos_config.py:193]: two blank lines between functions please :)

File: network/vyos/vyos_config.py [hunk 8]
```diff
@@ -133,74 +189,141 @@ def config_to_commands(config):
...
+    return contents
 
-def do_lines(module, result):
-    commands = module.params['lines']
-    result.update(load_config(module, commands))
+def get_candidate(module):
+    contents = module.params['src'] or module.params['lines']
+    if module.params['lines']:
+        contents = '\n'.join(contents)
+    return config_to_commands(contents)
```
Comment (gundalow) [inline network/vyos/vyos_config.py:211]: two blank lines between functions please :)

File: network/vyos/vyos_config.py [hunk 9]
```diff
@@ -133,74 +189,141 @@ def config_to_commands(config):
...
+                updates.append(line)
+            else:
+                item = re.sub(r'delete', 'set', item)
+                for entry in config:
+                    if entry.startswith(item) and not line in visited:
+                        updates.append(line)
+                        visited.add(line)
+
+    return list(updates)
+
```
Comment (gundalow) [inline network/vyos/vyos_config.py:238]: two blank lines between functions please :)

File: network/vyos/vyos_config.py [hunk 10]
```diff
@@ -133,74 +189,141 @@ def config_to_commands(config):
...
+def sanitize_config(config, result):
+    result['removed'] = list()
+    for regex in CONFIG_FILTERS:
+        for index, line in enumerate(list(config)):
+            if regex.search(line):
+                result['removed'].append(line)
+                del config[index]
+
```
Comment (gundalow) [inline network/vyos/vyos_config.py:246]: two blank lines between functions please :)

**Comment by gundalow** [comment]
needs_revision

**Comment by gregdek** [comment]
Thanks @privateip for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

File: network/vyos/vyos_config.py (hunk 2 above)
Comment (privateip) [reply network/vyos/vyos_config.py:172]: this should be "common sense" but i included the warning as a just in case.

**Comment by privateip** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @privateip. To the current maintainers, @privateip, @gundalow please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gundalow** [comment]
Shipit

**Pull request merged**